File: app/src/lib/events.ts
/**
 * Type-safe event system for inter-window communication.
 *
 * Events are broadcast to all windows via Tauri's event system.
 * This module provides type-safe wrappers around emit/listen.
 *
 * IMPORTANT: Event names and payload types must match the Rust side.
 * See: src-tauri/src/events.rs
 */

import { emit, listen, type UnlistenFn } from "@tauri-apps/api/event";
import type { ActiveAppContextSnapshot } from "./activeAppContext";

// =============================================================================
// Event Names - Must match src-tauri/src/events.rs
// =============================================================================

export const AppEvents = {
	// Rust → All: Hotkey triggers
	recordingStart: "recording-start",
	recordingStop: "recording-stop",
	prepareRecording: "prepare-recording",

	// Rust → All: Config sync notifications
	configResponse: "config-response",

	// Rust → Overlay: Disconnect request on app quit
	requestDisconnect: "request-disconnect",

	// Main → Overlay: Settings changed, refetch needed
	settingsChanged: "settings-changed",

	// Main → Overlay: Request reconnection
	reconnectRequest: "request-reconnect",

	// Overlay → Main: Connection state updates
	connectionState: "connection-state-changed",

	// Overlay → Main: Reconnection progress
	reconnectStarted: "reconnect-started",
	reconnectResult: "reconnect-result",

	// Rust → All: History changed
	historyChanged: "history-changed",

	// Overlay → Main: LLM error notification
	llmError: "llm-error",

	// Main → Overlay: Provider change request (pessimistic updates)
	providerChangeRequest: "provider-change-request",

	// Rust → All: Active app context updates
	activeAppContextChanged: "active-app-context-changed",
} as const;

// =============================================================================
// Event Payloads - Must match src-tauri/src/events.rs
// =============================================================================

export type ConnectionState =
	| "disconnected"
	| "connecting"
	| "reconnecting"
	| "idle"
	| "recording"
	| "processing";

/**
 * Known config setting names.
 * Unknown strings are still allowed for forward compatibility.
 */
export const KNOWN_CONFIG_SETTING_NAMES = [
	"stt-provider",
	"llm-provider",
	"prompt-sections",
	"stt-timeout",
] as const;

export type KnownConfigSettingName =
	(typeof KNOWN_CONFIG_SETTING_NAMES)[number];
export type ConfigSettingName = KnownConfigSettingName | (string & {});

/**
 * Config response for successful updates.
 * Value is parsed at runtime using Zod schemas in tauri.ts.
 */
export type ConfigUpdatedResponse = {
	type: "config-updated";
	setting: ConfigSettingName;
	value: unknown;
};

/**
 * Config error response.
 */
export type ConfigErrorResponse = {
	type: "config-error";
	setting: ConfigSettingName;
	error: string;
};

export type ConfigResponse = ConfigUpdatedResponse | ConfigErrorResponse;

export interface LLMErrorPayload {
	message: string; // Full error message for toast
	fatal: boolean;
}

export interface ProviderChangeRequestPayload {
	providerType: "stt" | "llm";
	value: string;
}

export interface EventPayloads {
	[AppEvents.recordingStart]: undefined;
	[AppEvents.recordingStop]: undefined;
	[AppEvents.prepareRecording]: undefined;
	[AppEvents.configResponse]: ConfigResponse;
	[AppEvents.requestDisconnect]: undefined;
	[AppEvents.settingsChanged]: undefined;
	[AppEvents.reconnectRequest]: undefined;
	[AppEvents.connectionState]: { state: ConnectionState };
	[AppEvents.reconnectStarted]: undefined;
	[AppEvents.reconnectResult]: { success: boolean; error?: string };
	[AppEvents.historyChanged]: undefined;
	[AppEvents.llmError]: LLMErrorPayload;
	[AppEvents.providerChangeRequest]: ProviderChangeRequestPayload;
	[AppEvents.activeAppContextChanged]: ActiveAppContextSnapshot;
}

// =============================================================================
// Type-safe emit/listen functions
// =============================================================================

/**
 * Emit an event with type-safe payload.
 */
export function emitEvent<K extends keyof EventPayloads>(
	event: K,
	...args: EventPayloads[K] extends undefined ? [] : [EventPayloads[K]]
): Promise<void> {
	return emit(event, args[0] ?? {});
}

/**
 * Listen for an event with type-safe callback.
 */
export function listenEvent<K extends keyof EventPayloads>(
	event: K,
	callback: (payload: EventPayloads[K]) => void,
): Promise<UnlistenFn> {
	return listen<EventPayloads[K]>(event, (eventPayload) =>
		callback(eventPayload.payload),
	);
}

File: app/src-tauri/src/events.rs
//! Type-safe event system for inter-window communication.
//!
//! Events are broadcast to all windows via Tauri's event system.
//! This module provides constants and types for event names and payloads.
//!
//! IMPORTANT: Event names and payload types must match the TypeScript side.
//! See: src/lib/events.ts

use serde::Serialize;

// =============================================================================
// Event Names - Must match src/lib/events.ts
// =============================================================================

/// Type-safe event names for Tauri event emission.
/// Use `EventName::*.as_str()` when calling `app.emit()`.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum EventName {
    /// Rust → All: Recording started
    RecordingStart,
    /// Rust → All: Recording stopped
    RecordingStop,
    /// Rust → All: Prepare for recording (mic warmup)
    PrepareRecording,
    /// Rust → All: Config sync response
    ConfigResponse,
    /// Rust → Overlay: Disconnect request on app quit
    RequestDisconnect,
    /// Main → Overlay: Settings changed, refetch needed
    SettingsChanged,
    /// Main → Overlay: Request reconnection
    ReconnectRequest,
    /// Overlay → Main: Connection state updates
    ConnectionState,
    /// Overlay → Main: Reconnection started
    ReconnectStarted,
    /// Overlay → Main: Reconnection result
    ReconnectResult,
    /// Rust → All: History changed
    HistoryChanged,
    /// Rust → Overlay: Native audio data from mic capture
    NativeAudioData,
    /// Rust → All: Active app context updates
    ActiveAppContextChanged,
}

impl EventName {
    /// Returns the string representation for Tauri event emission.
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::RecordingStart => "recording-start",
            Self::RecordingStop => "recording-stop",
            Self::PrepareRecording => "prepare-recording",
            Self::ConfigResponse => "config-response",
            Self::RequestDisconnect => "request-disconnect",
            Self::SettingsChanged => "settings-changed",
            Self::ReconnectRequest => "request-reconnect",
            Self::ConnectionState => "connection-state-changed",
            Self::ReconnectStarted => "reconnect-started",
            Self::ReconnectResult => "reconnect-result",
            Self::HistoryChanged => "history-changed",
            Self::NativeAudioData => "native-audio-data",
            Self::ActiveAppContextChanged => "active-app-context-changed",
        }
    }
}

// =============================================================================
// Config Setting Names - Must match src/lib/events.ts
// =============================================================================

/// Type-safe config setting names for config sync responses.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum ConfigSetting {
    PromptSections,
    SttTimeout,
    SttProvider,
    LlmProvider,
}

impl ConfigSetting {
    /// Returns the string representation for config response payloads.
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::PromptSections => "prompt-sections",
            Self::SttTimeout => "stt-timeout",
            Self::SttProvider => "stt-provider",
            Self::LlmProvider => "llm-provider",
        }
    }
}

// =============================================================================
// Event Payloads
// =============================================================================

#[derive(Debug, Clone, Serialize)]
#[serde(tag = "type", rename_all = "kebab-case")]
pub enum ConfigResponse<T: Serialize> {
    #[serde(rename = "config-updated")]
    Updated { setting: ConfigSetting, value: T },
    #[serde(rename = "config-error")]
    Error {
        setting: ConfigSetting,
        error: String,
    },
}

impl<T: Serialize> ConfigResponse<T> {
    pub fn updated(setting: ConfigSetting, value: T) -> Self {
        Self::Updated { setting, value }
    }

    pub fn error(setting: ConfigSetting, error: impl ToString) -> ConfigResponse<()> {
        ConfigResponse::Error {
            setting,
            error: error.to_string(),
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct ConnectionStatePayload {
    pub state: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct ReconnectResultPayload {
    pub success: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}
